tests/catch/stress/stream/streamEnqueue.cc

```diff
@@ -0,0 +1,227 @@
+  HIP_CHECK(
+      hipMemcpy(hPtr.get(), dPtr, sizeof(unsigned long long) * hwThreads, hipMemcpyDeviceToHost));
+
+  // Validate that CPU and GPU has the same results
+  for (size_t i = 0; i < hwThreads; i++) {
+    INFO("Check for Index " << i);
+    REQUIRE(hostData[i].data.load() == hPtr[i]);
+  }
+}
+
+__global__ void doOperation(int* dPtr, size_t size, int val) {
+  auto i = threadIdx.x;
+  atomicAdd(dPtr + i, val);
+}
+
+// Allocate mulitple stream for same device.
+// Same device stream operate on same memory
+TEST_CASE("Stress_StreamEnqueue_DifferentThreads_MultiGPU") {
+  int deviceCount{0};
+  HIP_CHECK(hipGetDeviceCount(&deviceCount));
+  REQUIRE(deviceCount > 0);
+
+  // Skip the test if devices less than 2
+  if (deviceCount <= 1) {
+    HIP_SKIP_TEST("Skipping because devices <= 1");
+    return;
+  }
+
+  constexpr size_t streamPerGPU{3};  // Stream per gpu
+
+  std::vector<hipStream_t> streamPool{};
+  streamPool.reserve(deviceCount * streamPerGPU);
+
+  std::map<hipStream_t, int*> streamToDeviceMemory;           // Map of stream and device memory
+  std::map<hipStream_t, AtomicWrap<int>> streamToHostMemory;  // Map of stream and host result
+  std::map<hipStream_t, size_t> streamToDeviceIndex;  // Map of stream and device it was created on
+
+  constexpr size_t size = 1024;
+
+  for (size_t i = 0; i < deviceCount; i++) {
+    HIP_CHECK(hipSetDevice(i));
+
+    for (size_t j = 0; j < streamPerGPU; j++) {
+      hipStream_t stream{nullptr};
+      HIP_CHECK(hipStreamCreate(&stream));
+      REQUIRE(stream != nullptr);
+      streamPool.push_back(stream);
+
+      int* dPtr{nullptr};
+      HIP_CHECK(hipMalloc(&dPtr, sizeof(int) * size));
+      REQUIRE(dPtr != nullptr);
+      HIP_CHECK(hipMemset(dPtr, 0, sizeof(int) * size));
+      HIP_CHECK(hipDeviceSynchronize());    // Wait till Memset is done
+      streamToDeviceMemory[stream] = dPtr;  // All streams work on exclusive memory
+
+      streamToHostMemory[stream] = AtomicWrap<int>(0);  // CPU result
+
+      streamToDeviceIndex[stream] = i;  // Capture device id for stream
+    }
+  }
+
+  constexpr size_t maxVal = 5;
+  constexpr size_t maxWorkPerThread = 10000;
+
+  // Boiler plate code to generate a random number
+  std::random_device device;
+  std::mt19937 engine(device());
+
+  std::uniform_int_distribution<std::mt19937::result_type> genVal(-maxVal, maxVal);
+  std::uniform_int_distribution<std::mt19937::result_type> genStream(0, streamPool.size() - 1);
+
+#if HT_NVIDIA
+  std::mutex ness;  // On nvidia, current device needs to match stream's device
+#endif
+
+  auto enqueueKernelThread = [&]() {
+    for (size_t i = 0; i < maxWorkPerThread; i++) {
+#if HT_NVIDIA
+      std::unique_lock<std::mutex> lock(ness);  // Lock on creation
+#endif
+      hipStream_t stream = streamPool[genStream(engine)];  // Get a random stream
+
+      hipSetDevice(streamToDeviceIndex[stream]);  // Cant do HIP_CHECK on this, catch2 macros and
```

Comment:
you can still check for hipError_t returned value

Reply:
done